orttraining/orttraining/python/training/onnxblock/_graph_utils.py

```diff
@@ -0,0 +1,205 @@
+    initializers = []
+    for initializer in model.graph.initializer:
+        if not initializer.name[0].isdigit():
+            # Move only those initializers as inputs that are not local
+            # to the onnx model. i.e. initializers that are model parameters.
+            # These are tpically those initializers without any number prefixed
+            # to their names.
+            graph_inputs.append(
+                onnx.helper.make_tensor_value_info(
+                    initializer.name, initializer.data_type, initializer.dims
+                )
+            )
+            if initializer.name not in user_args_not_requiring_grad:
+                all_args_requiring_gradient.add(initializer.name)
+        else:
+            # All other initializers stay where they were.
+            initializers.append(initializer)
+
+    # Update the initializers in the graph
+    del model.graph.initializer[:]
+    model.graph.initializer.extend(initializers)
+
+    # At this point, we have the eval model
+    accessor.eval_model = copy.deepcopy(model)
+
+    # Any graph input that requires gradient, should have been already added to
+    # args_requiring_grad. So, add these arguments to set of arguments
+    # whose gradient should be built.
+    for argument_name in user_args_requiring_grad:
+        all_args_requiring_gradient.add(argument_name)
+
+    # Assumption is that the first graph output is the loss output
+    if isinstance(output_names, str):
+        output_names = [output_names]
+    builder = GradientGraphBuilder(
+        model.SerializeToString(),
+        set(output_names),
+        all_args_requiring_gradient,
+        output_names[0],
+    )
+    builder.build()
+    gradient_model = onnx.load_from_string(builder.get_model())
+
+    # copy the gradient graph into the user's graph
+    model.graph.CopyFrom(gradient_model.graph)
+    del model.opset_import[:]
+    model.opset_import.extend(gradient_model.opset_import)
+
+    return all_args_requiring_gradient
+
+
+def build_gradient_accumulation_graph(grad_model, all_args_requiring_gradient_names):
+    """Builds gradient accumulation nodes on top of a training model.
+
+    Adds an InPlaceAccumulator node for every gradient so that the gradients
+    are accumulated in a gradient buffer (which is an input to InPlaceAccumulator).
+
+    For example, if there is a gradient in the graph called fc1.weight_grad,
+    an InPlaceAccumulator will be added for that gradient whose input will
+    be a graph input (fc1.weight_grad.accumulation.buffer) and the newly
+    computed gradient (fc1.weight_grad).
+
+    gradient.accumulation.buffer        gradient
+        |         |                         |
+        Ʌ         v                         v
+        |         |_________________________|
+        |                      |
+        |               InPlaceAccumulator
+        |                      |
+        |                      v
+        |______________________|
```

Comment:
I am thinking we should distinguish the first grad or the subsequent grad (to be accumulated) in our Module TrainStep implementation. That means, before session.Run(), we know whether the param.grad is empty or not. 
1. If it is empty, then we should fetch the results of the gradient (instead of InPlaceAccumulator);
2. Otherwise, we should fetch the results of the InPlaceAccumulator.

So I would suggest here, we put both the grad out name and the grad accumulation out name in Graph's outputs.

Reply:
Oh. Should we not allocate the gradient buffer before calling sess.Run()? That way, we can always fetch the gradient accumulation output. What is the downside of pre-allocating the gradient buffer?

Reply:
@ashbhandare can you confirm the expectation from the `Module::TrainStep` method? Does it expect the gradients as well as the accumulated gradients as output?